{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.7.9 64-bit ('SMWKG': conda)"
  },
  "interpreter": {
   "hash": "07b9714c10f0fa27e0393fb8b9b296c2ec1c2c727f2c6db093e91a70e66d2de2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "# imports\n",
    "from owlready2 import *\n",
    "import types\n",
    "import gzip\n",
    "import pandas as pd\n",
    "from rdflib import *\n",
    "import owlrl"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "# owlready2 functions\n",
    "\n",
    "# get Classes\n",
    "def getClasses(onto):        \n",
    "    return onto.classes()\n",
    "\n",
    "# get Data Properties    \n",
    "def getDataProperties(onto):        \n",
    "    return onto.data_properties()\n",
    "\n",
    "# get Objecct Properties       \n",
    "def getObjectProperties(onto):        \n",
    "    return onto.object_properties()\n",
    "\n",
    "# get Individuals   \n",
    "def getIndividuals(onto):        \n",
    "    return onto.individuals()\n",
    "\n",
    "# load ontology\n",
    "def loadOntology(urionto): \n",
    "    onto = get_ontology(urionto).load()\n",
    "    return onto\n",
    "\n",
    "# function to return labels for GO entities\n",
    "def my_rendering(entity):\n",
    "    return entity.label.first()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "# Load mappings from STRING ID to UNIPROT Protein ID for lookups / this code is reworked from OPA2Vec repository\n",
    "\n",
    "org_id = '9606' # Change to 9606 for Human\n",
    "\n",
    "# preprocess GO annotations\n",
    "mapping = {}\n",
    "source = {'4932': 'SGD_ID', '9606': 'Ensembl_UniProt_AC'} # mapping source\n",
    "\n",
    "with gzip.open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/STRINGDB/{org_id}.protein.aliases.v11.5.txt.gz', 'rt', encoding='utf-8') as f:\n",
    "    next(f) # Skip header\n",
    "    for line in f:\n",
    "        string_id, p_id, sources = line.strip().split('\\t')\n",
    "        if source[org_id] not in sources.split():\n",
    "            continue\n",
    "        if p_id not in mapping:\n",
    "            mapping[p_id] = set()\n",
    "        mapping[p_id].add(string_id)\n",
    "print('Loaded mappings', len(mapping))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Loaded mappings 83972\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "source": [
    "# load GO functions for each protein with mapping # code is reworked from OPA2Vec Repository\n",
    "gaf_files = {'4932': 'sgd.gaf.gz', '9606': 'goa_human.gaf.gz'}\n",
    "annotations = set()\n",
    "with gzip.open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/GeneAnnotationFile/{gaf_files[org_id]}', 'rt', encoding='utf-8') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('!'): # Skip header\n",
    "            continue\n",
    "        it = line.strip().split('\\t')\n",
    "        p_id = it[1]\n",
    "        go_id = it[4]\n",
    "        relation_id = it[3]\n",
    "        object_name = it[9]\n",
    "        if it[6] == 'IEA' or it[6] == 'ND': # Ignore predicted or no data annotations\n",
    "            continue\n",
    "        if p_id not in mapping: # Not in StringDB\n",
    "            continue\n",
    "        s_ids = mapping[p_id]\n",
    "        for s_id in s_ids:\n",
    "            annotations.add((s_id, go_id, object_name, relation_id))\n",
    "print('Number of annotations:', len(annotations))\n",
    "#print('Number of unique proteins:', )\n",
    "\n",
    "# Save annotations\n",
    "with open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/STRINGDB/{org_id}.proteinfunctions.txt', 'w') as f:\n",
    "    for s_id, go_id, object_name, relation_id in annotations:\n",
    "        f.write(f'{s_id}\\t{s_id}\\t{go_id}\\t{object_name}\\t{relation_id}\\n')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Number of annotations: 209463\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "with gzip.open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/GeneAnnotationFile/{gaf_files[org_id]}', 'rt', encoding='utf-8') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('!'): # Skip header\n",
    "            continue\n",
    "        it = line.strip().split('\\t')\n",
    "        print(it)\n",
    "        break"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['UniProtKB', 'A0A024RBG1', 'NUDT4B', 'enables', 'GO:0003723', 'GO_REF:0000043', 'IEA', 'UniProtKB-KW:KW-0694', 'F', 'Diphosphoinositol polyphosphate phosphohydrolase NUDT4B', 'NUDT4B', 'protein', 'taxon:9606', '20210612', 'UniProt']\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "# generate protein-proten interactions\n",
    "\n",
    "interactions = {}\n",
    "data = []\n",
    "with gzip.open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/STRINGDB/{org_id}.protein.links.v11.5.txt.gz', 'rt') as f:\n",
    "    next(f) # Skip header\n",
    "    for line in f:\n",
    "        p1, p2, score = line.strip().split()\n",
    "        if float(score) < 700: # Filter high confidence interactions\n",
    "            continue\n",
    "        if p1 not in interactions:\n",
    "            interactions[p1] = set()\n",
    "        if p2 not in interactions:\n",
    "            interactions[p2] = set()\n",
    "        if p2 not in interactions[p1]:\n",
    "            interactions[p1].add(p2)\n",
    "            interactions[p2].add(p1)\n",
    "            data.append((p1, p2))\n",
    "\n",
    "print('Total number of interactions:', len(data))\n",
    "print('Total number of proteins:', len(interactions.keys()))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Total number of interactions: 5969249\n",
      "Total number of proteins: 19385\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "len(data)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "5969249"
      ]
     },
     "metadata": {},
     "execution_count": 12
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# generate negative protein-protein interactions # OPA2Vec methodology and code\n",
    "import random\n",
    "proteins =set ()\n",
    "negatives = []\n",
    "for (p1,p2) in data:\n",
    "        proteins.add(p1)\n",
    "        proteins.add(p2)\n",
    "while len(negatives)< (len(data)/3):\n",
    "        s = random.sample(proteins, 2)\n",
    "        prot1= s[0]\n",
    "        prot2= s[1]\n",
    "        if (prot1,prot2) in negatives or (prot2,prot1) in negatives :\n",
    "                 continue\n",
    "        if prot1 not in interactions[prot2]:\n",
    "                 negatives.append((prot1, prot2))\n",
    "                 \n",
    "# Save negatives data\n",
    "with open(f'/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/STRINGDB/{org_id}.negative_interactions.txt', 'w') as f:\n",
    "    for p1 , p2 in negatives:\n",
    "        f.write(f'{p1}\\t{p2}\\t\\n')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# read protein links\n",
    "df_p_pairs = pd.read_csv('/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/9606.protein.links.v11.5.txt', sep='\\s', engine='python')\n",
    "\n",
    "# positive pairs: all scores above 0\n",
    "df_positive_pairs = df_p_pairs[df_p_pairs['combined_score'] > 0]\n",
    "# text preprocessing\n",
    "df_positive_pairs.loc[:,'protein1'] = df_positive_pairs.loc[:,'protein1'].str.replace('.','_')\n",
    "df_positive_pairs.loc[:,'protein2'] = df_positive_pairs.loc[:,'protein2'].str.replace('.','_')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "source": [
    "# inspect a protein\n",
    "df_positive_pairs.loc[df_positive_pairs['protein1'] == '9606_ENSP00000393963']"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9676792</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000265148</td>\n",
       "      <td>854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9676794</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000355090</td>\n",
       "      <td>710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9676807</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000274026</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9676812</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000404833</td>\n",
       "      <td>708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9676814</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000442068</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9677969</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000363794</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9677975</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000480987</td>\n",
       "      <td>808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9677989</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000362146</td>\n",
       "      <td>946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9678014</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000357643</td>\n",
       "      <td>833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9678041</th>\n",
       "      <td>9606_ENSP00000393963</td>\n",
       "      <td>9606_ENSP00000301633</td>\n",
       "      <td>837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     protein1              protein2  combined_score\n",
       "9676792  9606_ENSP00000393963  9606_ENSP00000265148             854\n",
       "9676794  9606_ENSP00000393963  9606_ENSP00000355090             710\n",
       "9676807  9606_ENSP00000393963  9606_ENSP00000274026             906\n",
       "9676812  9606_ENSP00000393963  9606_ENSP00000404833             708\n",
       "9676814  9606_ENSP00000393963  9606_ENSP00000442068             898\n",
       "...                       ...                   ...             ...\n",
       "9677969  9606_ENSP00000393963  9606_ENSP00000363794             784\n",
       "9677975  9606_ENSP00000393963  9606_ENSP00000480987             808\n",
       "9677989  9606_ENSP00000393963  9606_ENSP00000362146             946\n",
       "9678014  9606_ENSP00000393963  9606_ENSP00000357643             833\n",
       "9678041  9606_ENSP00000393963  9606_ENSP00000301633             837\n",
       "\n",
       "[77 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "execution_count": 12
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# number of unique proteins\n",
    "len(pd.unique(df_positive_pairs[['protein1','protein2']].values.ravel('K')))"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "16795"
      ]
     },
     "metadata": {},
     "execution_count": 13
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# make dataframe for annotations\n",
    "df_protein_functions = pd.DataFrame(annotations).rename(columns={ 0:'Protein_String_ID', 1:'GO_Term', 2:'Protein_Label', 3:\"Object_Property\"})\n",
    "# replace GO:0030314 with GO_0030314\n",
    "df_protein_functions['GO_Term'] = df_protein_functions['GO_Term'].replace(':','_', regex=True)\n",
    "df_protein_functions['Protein_String_ID'] = df_protein_functions['Protein_String_ID'].str.replace('.','_', regex=True)\n",
    "df_protein_functions['Object_Property'] = df_protein_functions['Object_Property'].astype(str).str.replace('NOT|','_not_', regex=False)"
   ],
   "outputs": [],
   "metadata": {
    "tags": []
   }
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "source": [
    "# inspect protein functions from Gene Annotation Files\n",
    "df_protein_functions"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein_String_ID</th>\n",
       "      <th>GO_Term</th>\n",
       "      <th>Protein_Label</th>\n",
       "      <th>Object_Property</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606_ENSP00000323421</td>\n",
       "      <td>GO_0008278</td>\n",
       "      <td>Structural maintenance of chromosomes protein 1A</td>\n",
       "      <td>part_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606_ENSP00000254436</td>\n",
       "      <td>GO_0010508</td>\n",
       "      <td>E3 ubiquitin-protein ligase TRIM21</td>\n",
       "      <td>involved_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606_ENSP00000255078</td>\n",
       "      <td>GO_0005524</td>\n",
       "      <td>DNA-binding protein SMUBP-2</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606_ENSP00000294740</td>\n",
       "      <td>GO_0005654</td>\n",
       "      <td>Zinc finger protein 281</td>\n",
       "      <td>located_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606_ENSP00000370473</td>\n",
       "      <td>GO_0031994</td>\n",
       "      <td>Insulin-like growth factor-binding protein 3</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209458</th>\n",
       "      <td>9606_ENSP00000335203</td>\n",
       "      <td>GO_0042030</td>\n",
       "      <td>ATPase inhibitor, mitochondrial</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209459</th>\n",
       "      <td>9606_ENSP00000237596</td>\n",
       "      <td>GO_0072208</td>\n",
       "      <td>Polycystin-2</td>\n",
       "      <td>involved_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209460</th>\n",
       "      <td>9606_ENSP00000265857</td>\n",
       "      <td>GO_0071816</td>\n",
       "      <td>Golgi to ER traffic protein 4 homolog</td>\n",
       "      <td>involved_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209461</th>\n",
       "      <td>9606_ENSP00000358921</td>\n",
       "      <td>GO_0005813</td>\n",
       "      <td>Alpha-centractin</td>\n",
       "      <td>located_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209462</th>\n",
       "      <td>9606_ENSP00000256460</td>\n",
       "      <td>GO_0051835</td>\n",
       "      <td>Calcium/calmodulin-dependent protein kinase ty...</td>\n",
       "      <td>involved_in</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>209463 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Protein_String_ID     GO_Term  \\\n",
       "0       9606_ENSP00000323421  GO_0008278   \n",
       "1       9606_ENSP00000254436  GO_0010508   \n",
       "2       9606_ENSP00000255078  GO_0005524   \n",
       "3       9606_ENSP00000294740  GO_0005654   \n",
       "4       9606_ENSP00000370473  GO_0031994   \n",
       "...                      ...         ...   \n",
       "209458  9606_ENSP00000335203  GO_0042030   \n",
       "209459  9606_ENSP00000237596  GO_0072208   \n",
       "209460  9606_ENSP00000265857  GO_0071816   \n",
       "209461  9606_ENSP00000358921  GO_0005813   \n",
       "209462  9606_ENSP00000256460  GO_0051835   \n",
       "\n",
       "                                            Protein_Label Object_Property  \n",
       "0        Structural maintenance of chromosomes protein 1A         part_of  \n",
       "1                      E3 ubiquitin-protein ligase TRIM21     involved_in  \n",
       "2                             DNA-binding protein SMUBP-2         enables  \n",
       "3                                 Zinc finger protein 281      located_in  \n",
       "4            Insulin-like growth factor-binding protein 3         enables  \n",
       "...                                                   ...             ...  \n",
       "209458                    ATPase inhibitor, mitochondrial         enables  \n",
       "209459                                       Polycystin-2     involved_in  \n",
       "209460              Golgi to ER traffic protein 4 homolog     involved_in  \n",
       "209461                                   Alpha-centractin      located_in  \n",
       "209462  Calcium/calmodulin-dependent protein kinase ty...     involved_in  \n",
       "\n",
       "[209463 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "execution_count": 16
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "source": [
    "# inspect all NOT relations \n",
    "for i in df_protein_functions['Object_Property']:\n",
    "        if 'not' in i :\n",
    "            print(i)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_contributes_to\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_is_active_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_contributes_to\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_contributes_to\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_acts_upstream_of_or_within\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_acts_upstream_of_or_within\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_is_active_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_acts_upstream_of_or_within\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_colocalizes_with\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_colocalizes_with\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_contributes_to\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_contributes_to\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_part_of\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_acts_upstream_of_or_within\n",
      "_not_located_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_is_active_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_located_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_colocalizes_with\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_enables\n",
      "_not_part_of\n",
      "_not_enables\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_involved_in\n",
      "_not_enables\n",
      "_not_enables\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Ontology Engineering"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# load relations ontology\n",
    "urionto_RO = '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/ro.xml'\n",
    "onto_RO = get_ontology(urionto_RO).load()\n",
    "# load GO ontology\n",
    "urionto_GO = '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_data/go.owl'\n",
    "onto_GO = get_ontology(urionto_GO).load() "
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# get list of object properties from relations ontology\n",
    "properties_RO = []\n",
    "for prop_ro in getObjectProperties(onto_RO):\n",
    "    properties_RO.append(prop_ro.iri)\n",
    "\n",
    "# get list of object properties from GO\n",
    "properties_GO = []\n",
    "for prop_go in getObjectProperties(onto_GO):\n",
    "    properties_GO.append(prop_go.iri)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "source": [
    "# string match function for object property alignment\n",
    "g1 = Graph()\n",
    "for prop_go in properties_GO:\n",
    "    for prop_ro in properties_RO:\n",
    "        if prop_go == prop_ro:\n",
    "            g1.add((URIRef(prop_go), OWL.equivalentProperty, URIRef(prop_ro)))\n",
    "        else:\n",
    "            continue    \n",
    "\n",
    "#serialize\n",
    "g1.serialize( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/property_matches.xml',format = 'xml')        "
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "source": [
    "# parse relations ontology, gene ontology and equivalent properties to same graph\n",
    "g_all = Graph()\n",
    "g_all.parse( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/go.owl')\n",
    "g_all.parse( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/ro.xml' )\n",
    "g_all.parse('/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/property_matches.xml')\n",
    "g_all.serialize('/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/GO_RO_processed.xml',format = 'xml')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "source": [
    "# load ontology with GO and Relations Ontology parsed inlcuding equivalent properties\n",
    "uriOnto = '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/GO_RO_processed.xml'\n",
    "onto = loadOntology(uriOnto)\n",
    "# get namespace\n",
    "onto_GO = onto.get_namespace(\"http://purl.obolibrary.org/obo/\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "source": [
    "# get all properties\n",
    "properties_all = {}\n",
    "for prop_go in getObjectProperties(onto):\n",
    "    if len(prop_go.label) > 0:\n",
    "        properties_all[prop_go.label[0]] = prop_go\n",
    "\n",
    "# change key names replaces whitespace with an underscore\n",
    "for key,value in properties_all.items():\n",
    "    properties_all = { k.replace(' ', '_'): v for k, v in properties_all.items() }"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "source": [
    "# modify protein functions dataframe to avoid duplicate equivalent classes\n",
    "dict_prots = {}\n",
    "for go_term, property_name in zip(df_protein_functions.GO_Term,df_protein_functions.Object_Property):\n",
    "    property_name = property_name.replace(' ', '_')\n",
    "    dict_prots[go_term] = property_name\n",
    "\n",
    "#for key, value in dict_prots.items():\n",
    "#    dict_prots[key] = value.replace(' ','_').replace('NOT|','_not_')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "source": [
    "# map equivalences and modify GO with RDFLib & Owlready2\n",
    "G_equivalences = Graph()\n",
    "n_space = Namespace(\"http://purl.obolibrary.org/obo/\")\n",
    "\n",
    "with onto_GO:\n",
    "    # add new Protein class in correct namespace\n",
    "    class Protein(Thing):\n",
    "        pass\n",
    "    \n",
    "    # make new Protein Class subclass of Thing\n",
    "    superClass = [Thing]\n",
    "\n",
    "    # create new equivalent protein classes (i.e. obo.protein_involved_in_GO_0003132 owl:equivalentClass obo.Protein)\n",
    "    for cls in getClasses(onto):\n",
    "        if str(cls)[4:] in dict_prots.keys():\n",
    "            class_name = 'protein_' + dict_prots[str(cls)[4:]] + '_' + str(cls)[4:]      \n",
    "            \n",
    "            # make new protein class\n",
    "            new_protein_class = types.new_class(class_name, tuple(superClass))\n",
    "            \n",
    "            #use RDFLib instead of OWLREADY2 (slow) to write out equivalent classes\n",
    "            G_equivalences.add((URIRef(n_space + class_name), OWL.equivalentClass, URIRef(n_space + 'Protein')))\n",
    "            \n",
    "            # lookup relevant property, create new protein and generate existential restriction (retrived relation, some, go_term)\n",
    "            lookup_go = dict_prots[str(cls)[4:]] \n",
    "            if lookup_go in properties_all.keys():\n",
    "                new_protein_class.is_a.append(properties_all[lookup_go].some(cls)) \n",
    "    \n",
    "    # serialize equivalences      \n",
    "    G_equivalences.serialize('/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/equivalent_class.xml', format = 'xml')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "source": [
    "#save modified GO with Owlready2\n",
    "onto.save('/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/processed_ontology_human.xml', format = \"rdfxml\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "source": [
    "# load ontology\n",
    "uriOnto = '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/processed_ontology_human.xml'\n",
    "onto = loadOntology(uriOnto)\n",
    "\n",
    "# get namespace\n",
    "onto_GO = onto.get_namespace(\"http://purl.obolibrary.org/obo/\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# check restrictions work\n",
    "onto_GO.protein_involved_in_GO_0070370.is_a"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "source": [
    "# make dictionary of classes with URI\n",
    "GO_dictionary = {}\n",
    "with onto_GO:\n",
    "    for cls in getClasses(onto):\n",
    "        GO_dictionary[str(cls)] = cls.iri"
   ],
   "outputs": [],
   "metadata": {
    "tags": []
   }
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "source": [
    "# initialize empty graph\n",
    "g = Graph()\n",
    "\n",
    "t0 = time.time()\n",
    "n_space = Namespace(\"http://purl.obolibrary.org/obo/\")\n",
    "\n",
    "# associate proteins to new ontology via correct object property \n",
    "for index,row in df_protein_functions.iterrows():\n",
    "    class_name = 'protein_' + str(row['Object_Property']) + '_' + str(row['GO_Term'])\n",
    "    protein_label = row['Protein_Label']\n",
    "    protein_string_id = n_space + row['Protein_String_ID']\n",
    "    # add triples\n",
    "    g.add((URIRef(protein_string_id), RDF.type, URIRef(n_space + class_name)))\n",
    "    g.add((URIRef(protein_string_id), RDFS.label, Literal(protein_label)))\n",
    "    g.add((URIRef(protein_string_id), RDF.type, OWL.NamedIndividual))\n",
    "\n",
    "# save graph\n",
    "g.serialize(destination='/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/protein_instances_rdf.xml', format='xml')\n",
    "\n",
    "# duration\n",
    "t1 = time.time()\n",
    "total = t1-t0\n",
    "print(total)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "64.8437180519104\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "source": [
    "# initialize empty graph\n",
    "g = Graph()\n",
    "\n",
    "t0 = time.time()\n",
    "n_space = Namespace(\"http://purl.obolibrary.org/obo/\")\n",
    "\n",
    "# generate protein instances for GO terms. Because of prior equivalence mapping, proteins will also \n",
    "# be instances of equivalent classes (i.e. p1 rdf:type protein_located_in_GO_0000323, where obo.protein_located_in_GO_000323 owl:equivalentClass obo.Protein)\n",
    "for index,row in df_protein_functions.iterrows():\n",
    "    # get go_term affiliated with relevant protein\n",
    "    key = \"_\" + row['GO_Term']\n",
    "    # list comprehension to match GO term against dictionary\n",
    "    key_name = [k for k in GO_dictionary.keys() if key in k]\n",
    "    if len(key_name) > 0:\n",
    "        go_term = GO_dictionary[key_name[0]]\n",
    "        protein_string_id = n_space + row['Protein_String_ID']\n",
    "        protein_label = row['Protein_Label']\n",
    "        # add triple\n",
    "        g.add((URIRef(protein_string_id), RDF.type, URIRef(go_term)))\n",
    "        g.add((URIRef(protein_string_id), RDFS.label, Literal(protein_label)))\n",
    "        g.add((URIRef(protein_string_id), RDF.type, OWL.NamedIndividual))\n",
    "\n",
    "# save graph\n",
    "g.serialize(destination='/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/protein_instances_rdf.xml', format='xml')\n",
    "\n",
    "t1 = time.time()\n",
    "total = t1-t0\n",
    "print(total)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "1201.2345070838928\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "source": [
    "# load empty graph\n",
    "g1 = Graph()\n",
    "\n",
    "# parse processed GO (with RO object properties, and equivalent properties mapped) and protein instances of new equivalent classes\n",
    "g1.parse( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/processed_ontology_human.xml', format = 'xml')\n",
    "g1.parse( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/protein_instances_rdf.xml', format = 'xml')\n",
    "g1.parse( '/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/equivalent_class.xml', format = 'xml')\n",
    "\n",
    "# save modified protein ontology\n",
    "g1.serialize(destination='/Users/teissherman/Desktop/GitHub/IndividualProject/Project/data/Ontology_Data/preprocessed_all_data_xml.xml', format='xml')"
   ],
   "outputs": [],
   "metadata": {}
  }
 ]
}